src/classic/clvm_tools/stages/stage_2/module.rs

```diff
@@ -385,26 +397,110 @@ fn compile_mod_stage_1(
                         &mut namespace,
                         &mut functions,
                         &mut constants,
+                        &mut delayed_constants,
                         &mut macros,
                         run_program.clone()
                     )?;
                 }
 
-                let uncompiled_main = alist[alist.len() - 1];
-                m! {
-                    main_list <-
-                        enlist(
+                // For each delayed constant, drain it into the
+                // main constant pool.
+                let mut result_collection = CollectionResult {
+                    functions,
+                    constants,
+                    macros
+                };
+
+                let main_name_vec = MAIN_NAME.as_bytes().to_vec();
+
+                // Process delayed constants until we either can't advance or
+                // they're all done.
+
+                loop {
+                    if delayed_constants.is_empty() {
```

Comment:
I think I see why you do this. You don't know the order in which they need to be evaluated up-front, so you loop until they're all evaluated.

Do you have a test where you have cycles? And also other mixtures of orders.

Reply:
The tests include cycles and out of order sequence but I'll add more.

Reply:
I actually later made a general toposort algorithm and put it in util (second time writing it, for (assign ...) form in a different pr.  I considered backporting it here but thought it might be better to introduce it in a different pr.  If it seems like a good idea i'll pull that into a separate pr and remove the bespoke toposort here.

Reply:
Also added a more complex test.  Please tell me what you think of the above.